crates/txtx-core/src/runbook/location.rs

```diff
@@ -0,0 +1,355 @@
+        let mut col = 1;
+
+        for (i, ch) in self.source.char_indices() {
+            if i >= start {
+                break;
+            }
+            if ch == '\n' {
+                line += 1;
+                col = 1;
+            } else {
+                col += 1;
+            }
+        }
+
+        (line, col)
+    }
+
+    /// Convert an optional span to a location, returning a default if None
+    pub fn optional_span_to_location(
+        &self,
+        span: Option<&Range<usize>>,
+        file: String,
+    ) -> SourceLocation {
+        match span {
+            Some(s) => self.span_to_location(s, file),
+            None => SourceLocation::at_start(file),
+        }
+    }
+
+    /// Convert an optional span to position, returning (1, 1) if None
+    pub fn optional_span_to_position(&self, span: Option<&Range<usize>>) -> (usize, usize) {
+        span.map(|s| self.span_to_position(s)).unwrap_or((1, 1))
+    }
+}
+
+/// Context of where a reference or definition appears in the HCL structure
+#[derive(Debug, Clone, PartialEq, Eq)]
+pub enum BlockContext {
+    /// Inside an action block
+    Action(String),
+    /// Inside a variable block
+    Variable(String),
+    /// Inside a signer block
+    Signer(String),
+    /// Inside an output block
+    Output(String),
+    /// Inside a flow block
+    Flow(String),
+    /// Inside an addon block
+    Addon(String),
+    /// Unknown or top-level context
+    Unknown,
+}
+
+impl BlockContext {
+    /// Extract the name from the context if available
+    pub fn name(&self) -> Option<&str> {
+        match self {
+            BlockContext::Action(name)
+            | BlockContext::Variable(name)
+            | BlockContext::Signer(name)
+            | BlockContext::Output(name)
+            | BlockContext::Flow(name)
+            | BlockContext::Addon(name) => Some(name),
+            BlockContext::Unknown => None,
+        }
+    }
+
+    /// Get the block type as a string
+    pub fn block_type(&self) -> &str {
+        match self {
+            BlockContext::Action(_) => "action",
+            BlockContext::Variable(_) => "variable",
+            BlockContext::Signer(_) => "signer",
+            BlockContext::Output(_) => "output",
+            BlockContext::Flow(_) => "flow",
+            BlockContext::Addon(_) => "addon",
+            BlockContext::Unknown => "unknown",
+        }
+    }
+}
+
+/// Type of reference being tracked
+#[derive(Debug, Clone, Copy, PartialEq, Eq)]
+pub enum ReferenceType {
+    /// Reference to an input (input.*)
+    Input,
+    /// Reference to a variable (var.* or variable.*)
+    Variable,
+    /// Reference to an action (action.*)
+    Action,
+    /// Reference to a signer (signer.*)
+    Signer,
+    /// Reference to a flow input (flow.*)
+    FlowInput,
+    /// Reference to an output (output.*)
+    Output,
+}
+
+/// A reference to an input, variable, or other construct in the runbook
+#[derive(Debug, Clone)]
+pub struct InputReference {
```

Comment:
I don't think this struct is used anywhere and can be removed. (plus the test referencing it below)